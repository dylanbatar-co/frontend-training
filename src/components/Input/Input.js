import React, { useState } from 'react';
import TextField from 'emerald-ui/lib/TextField';
import Col from 'emerald-ui/lib/Col';

export const Input = ({
  onChange,
  value,
  name,
  label,
  fullCol,
  textarea,
  validator,
  ...otherProps
}) => {
  const [error, setError] = useState('');

  return (
    <Col xs={12} lg={textarea ? 12 : 6} md={textarea ? 12 : 6}>
      <label htmlFor={`${name}`}>{label}</label>
      <TextField
        id={`${name}`}
        required
        name={name}
        onBlur={() => setError(er => setError(validator(value, label)))}
        onChange={onChange}
        value={value}
        textarea={textarea}
        errorMessage={error}
        {...otherProps}
      />
    </Col>
  );
};
