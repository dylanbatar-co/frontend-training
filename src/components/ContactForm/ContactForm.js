import React, { useState } from 'react';

import Card from 'emerald-ui/lib/Card';
import Checkbox from 'emerald-ui/lib/Checkbox';
import Button from 'emerald-ui/lib/Button';

import Container from 'emerald-ui/lib/Container';
import Row from 'emerald-ui/lib/Row';
import Col from 'emerald-ui/lib/Row';

import { useForm } from '../../hooks/useForm';
import { Input } from '../Input/Input';
import { Modal } from '../Modal/Modal';

import {
  emailValidation,
  fieldRequired,
  phoneValidation
} from '../../utils/validations';

export const ContactForm = () => {
  const [
    { email, firstName, lastName, message, phone, news },
    handleInputChange,
    handleCheckbox
  ] = useForm({
    email: '',
    firstName: '',
    lastName: '',
    message: '',
    phone: '',
    news: false
  });

  const [showModal, setShowModal] = useState(false);
  const [isSubmit, setIsSubmit] = useState(false);

  const handlerSubmit = event => {
    event.preventDefault();
    if (
      Object.values({ email, firstName, lastName, message, phone }).includes('')
    ) {
      setIsSubmit(true);
      return;
    }
    setShowModal(true);
  };

  return (
    <form className="contact-form" onSubmit={handlerSubmit}>
      <Card className="pt-5 pb-5">
        <Container>
          <Row>
            <Input
              value={firstName}
              onChange={handleInputChange}
              label="First name"
              name="firstName"
              validator={fieldRequired}
              isSubmit={isSubmit}
            />
            <Input
              value={lastName}
              onChange={handleInputChange}
              label="Last name"
              name="lastName"
              validator={fieldRequired}
              isSubmit={isSubmit}
            />
          </Row>
          <Row>
            <Input
              value={email}
              onChange={handleInputChange}
              type="email"
              label="Email"
              name="email"
              validator={emailValidation}
              isSubmit={isSubmit}
            />
            <Input
              value={phone}
              onChange={handleInputChange}
              label="Phone number"
              name="phone"
              type="phone"
              validator={phoneValidation}
              isSubmit={isSubmit}
            />
          </Row>
          <Row>
            <Input
              value={message}
              onChange={handleInputChange}
              name="message"
              label="Message"
              textarea
              validator={fieldRequired}
              isSubmit={isSubmit}
            />
          </Row>
          <Checkbox
            checked={news}
            onChange={handleCheckbox}
            className="pl-0"
            name="news"
            label="Send me emails about breaking news and promotions."
          />
          <Col className="container-button mt-5">
            <Button type="submit" color="info">
              Submit form
            </Button>
          </Col>
        </Container>
      </Card>
      <>
        <Modal
          showProp={showModal}
          toggleModalProp={setShowModal}
          data={{ email, firstName, lastName, message, phone, news }}
        />
      </>
    </form>
  );
};
