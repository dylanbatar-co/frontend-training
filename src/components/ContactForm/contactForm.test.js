import React from 'react';
import { shallow } from 'enzyme';
import { Input } from '../Input/Input';
import { ContactForm } from './ContactForm';
import {
  emailValidation,
  fieldRequired,
  phoneValidation
} from '../../utils/validations';

describe('Contact form Tests', () => {
  const renderInput = args => {
    return shallow(<Input {...args} />).find('TextField');
  };

  it('Component snapshot', () => {
    const component = shallow(<ContactForm />);
    expect(component).toMatchSnapshot();
  });

  it('TextField is required', () => {
    const props = {
      name: 'firstName',
      label: 'First Name',
      value: '',
      onChange: jest.fn()
    };

    const TextField = renderInput(props);
    const { value } = TextField.props();
    expect(fieldRequired(value, props.label)).toEqual(
      `${props.label} is required`
    );
  });

  it('Email is valid (Format)', () => {
    const props = {
      name: 'email',
      value: '',
      onChange: jest.fn()
    };

    const TextField = renderInput(props);
    expect(emailValidation(TextField.prop('value'))).toEqual(
      'Enter a valid email'
    );

    expect(emailValidation('email@domain.com')).toBe(null);
  });

  it('Phone is valid (Format)', () => {
    const props = {
      name: 'phone',
      value: '',
      onChange: jest.fn()
    };
    const TextField = renderInput(props);
    expect(phoneValidation(TextField.prop('value'))).toEqual(
      'Enter a valid phone'
    );
  });

  it('Blur field', () => {
    const validatorFunct = jest.fn();
    const TextField = renderInput({ validator: validatorFunct });
    TextField.simulate('blur');
    expect(validatorFunct).toHaveBeenCalled();
  });

  it('Open Modal when form is submitted', () => {
    const component = shallow(<ContactForm />);
    component.simulate('submit', { preventDefault() {} });
    const modal = component.find('Modal');
    expect(modal.prop('showProp')).toBe(true);
  });
});
