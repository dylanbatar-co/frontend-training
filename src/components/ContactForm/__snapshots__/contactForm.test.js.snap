// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact form Tests Component snapshot 1`] = `
<form
  className="contact-form"
  onSubmit={[Function]}
>
  <Card
    className="pt-5 pb-5"
  >
    <Container>
      <Row>
        <Input
          label="First name"
          name="firstName"
          onChange={[Function]}
          validator={[Function]}
          value=""
        />
        <Input
          label="Last name"
          name="lastName"
          onChange={[Function]}
          validator={[Function]}
          value=""
        />
      </Row>
      <Row>
        <Input
          label="Email"
          name="email"
          onChange={[Function]}
          type="email"
          validator={[Function]}
          value=""
        />
        <Input
          label="Phone number"
          name="phone"
          onChange={[Function]}
          type="phone"
          validator={[Function]}
          value=""
        />
      </Row>
      <Row>
        <Input
          label="Message"
          name="message"
          onChange={[Function]}
          textarea={true}
          validator={[Function]}
          value=""
        />
      </Row>
      <Checkbox
        className="pl-0"
        label="Send me emails about breaking news and promotions."
      />
      <Row
        className="container-button mt-5"
      >
        <Button
          color="info"
          displayName="Button"
          shape="classic"
          size="xs"
          type="submit"
        >
          Submit form
        </Button>
      </Row>
    </Container>
  </Card>
  <Modal
    data={
      Object {
        "email": "",
        "firstName": "",
        "lastName": "",
        "message": "",
        "phone": "",
      }
    }
    showProp={false}
    toggleModalProp={[Function]}
  />
</form>
`;
